### PR TITLE
cmd/tailscaled: make tailscaled run off its own internal state.

### PR DESCRIPTION
**Pull request by danderson**

### CODE REVIEW

File: cmd/tailscale/tailscale.go [hunk 1]
```diff
@@ -45,34 +42,26 @@ func main() {
 		log.Printf("fixConsoleOutput: %v\n", err)
 	}
 
-	config := getopt.StringLong("config", 'f', "", "path to config file")
 	server := getopt.StringLong("server", 's', "https://login.tailscale.com", "URL to tailcontrol server")
 	nuroutes := getopt.BoolLong("no-single-routes", 'N', "disallow (non-subnet) routes to single nodes")
-	rroutes := getopt.BoolLong("remote-routes", 'R', "allow routing subnets to remote nodes")
-	droutes := getopt.BoolLong("default-routes", 'D', "allow default route on remote node")
+	routeall := getopt.BoolLong("remote-routes", 'R', "accept routes advertised by remote nodes")
+	nopf := getopt.BoolLong("no-packet-fitler", 'F', "disable packet filter")
```
Comment (bradfitz) [inline cmd/tailscale/tailscale.go]: typo in flag name (fitler)

File: cmd/tailscaled/tailscaled.go [hunk 2]
```diff
@@ -60,6 +64,7 @@ func main() {
 
 	opts := ipnserver.Options{
 		StatePath:          *statepath,
+		AutostartStateKey:  "root",
```
Comment (bradfitz) [inline cmd/tailscaled/tailscaled.go]: make "root" a named constant (unexported is fine) so you can clarify in attached docs that this is or is not related to the "root" Unix account. It might be worth a less ambiguous name if it's unrelated to user/group names. (I think it means "default" or "base" or "daemon" key?)

File: ipn/ipnserver/server.go [hunk 3]
```diff
@@ -25,11 +25,20 @@ import (
...
+	"tailscale.com/version"
 	"tailscale.com/wgengine"
 )
 
+// Options is the configuration of the Tailscale node agent.
 type Options struct {
-	StatePath          string
+	// StatePath is the path to the stored agent state.
+	StatePath string
+	// AutostartStateKey immediately starts the agent using the given
```
Comment (bradfitz) [inline ipn/ipnserver/server.go]: // AustartStateKey, if non-empty, immediately starts the ...

File: ipn/ipnserver/server.go [hunk 4]
```diff
@@ -25,11 +25,20 @@ import (
...
+	// StateKey. If empty, the agent stays idle and waits for a
+	// frontend to start it.
+	AutostartStateKey ipn.StateKey
+	// SurviveDisconnects specifies whether the agent should log out
```
Comment (bradfitz) [inline ipn/ipnserver/server.go]: That sounds like the opposite of the field name.

When using "whether", the text typically describes the "true" case. If that's not what you meant then say something like "specifies how the agent's connection to foo stays connected when the frontend disconnects. If true, foo. If false, bar."

File: ipn/ipnserver/server.go [hunk 5]
```diff
@@ -86,13 +101,18 @@ func Run(rctx context.Context, logf logger.Logf, logid string, opts Options, e w
...
-	go func() {
-		<-rctx.Done()
-		listen.Close()
-	}()
+	if opts.AutostartStateKey != "" {
+		cmd := &ipn.Command{
+			Version: version.LONG,
+			Start: &ipn.StartArgs{
+				Opts: ipn.Options{
+					ServerURL: "https://login.tailscale.com",
```
Comment (bradfitz) [inline ipn/ipnserver/server.go]: use a const

File: ipn/ipnserver/server.go [hunk 6]
```diff
@@ -86,13 +101,18 @@ func Run(rctx context.Context, logf logger.Logf, logid string, opts Options, e w
...
+					StateKey:  opts.AutostartStateKey,
+				},
+			},
+		}
+		bs.GotCommand(cmd)
```
Comment (bradfitz) [inline ipn/ipnserver/server.go]: no need for the cmd var? inline it here?

File: ipn/message.go [hunk 7]
```diff
@@ -31,9 +31,11 @@ type FakeExpireAfterArgs struct {
 	Duration time.Duration
 }
 
-// A command message sent to the server. Exactly one of these must be non-nil.
+// Command is a command message sent by a frontend to a backend.
```
Comment (bradfitz) [inline ipn/message.go]: clarify that it's JSON-encoded, since I haven't done a pass over all these to add explicit JSON struct tags.

File: cmd/tailscale/tailscale.go [hunk 8]
```diff
@@ -108,41 +98,22 @@ func main() {
 					fmt.Fprintf(os.Stderr, "\nTo authorize your machine, visit (as admin):\n\n\t%s/admin/machines\n\n", *server)
 				case ipn.Starting, ipn.Running:
 					// Done full authentication process
+					fmt.Fprintf(os.Stderr, "\ntailscaled is authenticated, nothing more to do.\n\n")
```
Comment (bradfitz) [inline cmd/tailscale/tailscale.go]: will this be spammy? If this is going to say this every time, at least trim it down to "Tailscaled authenticated." ?

File: cmd/tailscaled/tailscaled.go (hunk 2 above)
Comment (danderson) [reply cmd/tailscaled/tailscaled.go]: Used `_daemon` because that's very un-username-ish, so keeps open the option of per-user state down the road.

File: ipn/ipnserver/server.go (hunk 4 above)
Comment (danderson) [reply ipn/ipnserver/server.go]: Oops. Fixed.

File: cmd/tailscale/tailscale.go (hunk 8 above)
Comment (danderson) [reply cmd/tailscale/tailscale.go]: It'll print this once per run of the `tailscale` CLI, because it exits right after printing this. In this current (to be replaced) CLI paradigm, you run `tailscale` and it tells you one of "needs auth, go to this URL" or "no auth needed, all's well".

It's getting replaced with a new CLI soon.
